Matter_Engine/Query/Density.cs

```diff
@@ -0,0 +1,174 @@
+using BH.oM.Reflection.Attributes;
+using BH.oM.Physical.Elements;
+
+using BH.Engine.Base;
+using BH.oM.Physical.FramingProperties;
+using BH.oM.Physical.Materials;
+using BH.oM.Reflection;
+using BH.oM.Quantities.Attributes;
+
+namespace BH.Engine.Physical
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Gets the density of a Material by querying each of its IMaterialProperties for theirs." +
+                     "The density is either gotten from a method or a property with that name." + 
+                     "If the density is found on diffrent IMaterialProperties, the avarage is taken.")]
+        [Input("material", "The material to get the density of")]
+        [Input("type", "The kind of IMaterialProperties to cull the result by")]
+        [Output("density", "The density of the material, further info on how the value was accuired is recorded in the warning", typeof(Density))]
+        public static double Density(this Material material, Type type = null)
+        {
+            if (type == null)
+                type = typeof(IMaterialProperties);
+
+            List<double> densities = new List<double>();
+            List<string> warnings = new List<string>() { "Density report for " + material.Name + ":" };
+
+            foreach (IMaterialProperties mat in material.Properties.Where(x => type.IsAssignableFrom(x.GetType())))
+            {
+                Output<double, string> result = IDensityWithReport(mat, 293, 0.8);
+                if (!double.IsNaN(result.Item1))
+                {
+                    if (!string.IsNullOrWhiteSpace(result.Item2))
+                        warnings.Add("  " + result.Item2);
+                    densities.Add(result.Item1);
+                }
+            }
+
+            if (densities.Count == 0)
+            {
+                Reflection.Compute.RecordWarning("no density on any of the fragments of " + material.Name + " by type " + type.Name);
+                return 0;
+            }
+            if (densities.Count > 1)
+                warnings.Add("The density for " + material.Name + " is a average of multiple Fragments");
+
+            if (warnings.Count > 1)
+                Reflection.Compute.RecordWarning(string.Join(System.Environment.NewLine, warnings.ToArray()));
+
+            return densities.Average();
+        }
+
+        /***************************************************/
+
+        [Description("Gets the density of a IMaterialProperties. The density is either gotten from a method or a property with that name.")]
+        [Input("material", "The IMaterialProperties to get the density of")]
+        [Input("temprature", "The temprature to get the density at", typeof(Temperature))]
+        [Input("humidity", "The humidity to get the density at", typeof(Ratio))]
```

Comment:
would type out "relativeHumidity" instead of just "humidity" here to make it more clear